### ISSUE TITLE
Always print immediates as unsigned in the disassembly

### ISSUE THREAD

**Issue by pankkor** [cmuratori/computer_enhance#44]
I'm perhaps late to the party.

I have a problem with how Immediate are being printed out in the disassembly.

- https://github.com/cmuratori/computer_enhance/blob/main/perfaware/part1/listing_0045_challenge_register_movs.txt#L5

    Binary:
    ```
    BA 88 88
    ```
    Your reference sim86 disassembles it as:
    ```
    mov dx, 34952
    ```

- https://github.com/cmuratori/computer_enhance/blob/main/perfaware/part1/listing_0047_challenge_flags.txt#L12

    Binary
    ```
    81 C3 40 9C
    83 C1 A6
    ```
    Your reference sim86 disassembles it as:
    ```
    add cx, 40000
    add cx, -90
    ```
As we can't differentiate between positive and negative immediates in the binary, the last `-90` seems to be out of place. I would appreciate if it would be printed as unsigned 65446.

Immediate is printed as signed integer here:
https://github.com/cmuratori/computer_enhance/blob/main/perfaware/sim86/sim86_text.cpp#L128

And this is how it's read:
https://github.com/cmuratori/computer_enhance/blob/main/perfaware/sim86/sim86_decode.cpp#L62

When we read wide, 2 bytes are read and stored into lower word of 32-bit Result. No sign extension happens if the word has bit 15 set, hence this 16-bit immediate would be always printed as a positive 32-bit integer.
However in case of sign extension we sign extend all 32-bits of Result. This immediate could be printed out as a negative 32-bit integer.
I think it would be more appropriate to sign extend only up to a word boundary and always print as unsigned. What do you think?

**Comment by cmuratori** [comment]
I'm not averse to it printing the value as unsigned, but I also don't know what benefit it actually provides. If you wanted to actually be conscientious about printing signed or unsigned, I think you would probably want to have more logic in general - something where you look at the code and make an educated guess about whether the programmer would have used a signed or unsigned value.

So as it stands, I don't see any compelling reason to change how it disassembles, since the choice is arbitrary.

\- Casey

**Comment by pankkor** [comment]
I would not dare to make any educated guesses about programmer's intent. I merely want a consistent decoder output.
Printing signed or unsigned is indeed an arbitrary choice and I'm fine with any of it, as long as it's consistent, however now it's not consistent when we read 1 byte and sign extend it.

*tl;dr:* I would like all immediates to be printed as 16-bit values, signed or unsigned, I don't have any preference. I.e. I would be fine to have 0xFFFF immediate to always be printed as -1 *or* as 65535, no matter which decoder code path is taken. Now decoder can produce both 65535 and -1 in the same listing.

Let me elaborate.

Context:
In my decoder I store immediates as `u16` and print them them unsigned (I can also print them signed, I don't care). What I do is I go over your listings and make my simulator match your output exactly. So far it worked fine until I encountered that `-90` in the listing, which puzzled me. In all the previous listings all immediates were printed as unsigned 16-bit values and that `-90` seems to be out of place.

For instance [in the same listing](https://github.com/cmuratori/computer_enhance/blob/main/perfaware/part1/listing_0047_challenge_flags.txt#L11C1-L12) you have `add cx, 40000` and it's not printed as a negative 16-bit value. And in the next line you have `add cx, -90`.

Why does -90 printed as negative and 40000 printed as positive? They are both encoded as immediates which have no sign, why (and *how*) is there a difference? After decode immediates are stored as 32-bit value. And there are 2 different code paths that decode 2 bytes immediate and 1 byte immediate with sign extension. 1 byte immediate sign extension to 32-bit `Result`, happens here https://github.com/cmuratori/computer_enhance/blob/main/perfaware/sim86/sim86_decode.cpp#L62

1. When instruction has 1 byte immediate encoded, you read it and then sign extend it to all 32-bits of the `Result`. 
If you read `0xFF`, the sign extended `Result` would be `0xFFFF FFFF`. It would be later printed as a negative value -1.

2. However when instruction has 2 byte immediate encoded, you read those 2 bytes and store them in the lower word of the `Result` dword, no sign extensions happens. So if you read  byte 0xFF` and another byte `0xFF`, the `Result` would be `0x0000 FFFF`. It would be printed as a positive value 65535.

It's not enough to just change `printf("%d")` to `printf("%u")`, as 32-bit extended -1 would be printed as 4294967295, which would look peculiar for a 16-bit simulator.

That's why I proposed to sign extend only 16-bit of 32-bit `Result` when you read 1 byte immediate. If you read one `0xFF` byte and sign extend it, or read two `0xFF` bytes, the `Result` would be the same `0x0000 FFFF`.
This way we could pretend we deal with 16-bit immadiates. 

It's enough to change `Result = (s32)*(s8 *)&Result;` to `Result = (u16)*(s8 *)&Result;` to only sign extend up to 16-bits of the result.

Then _listing_0047_ would look like this:
```
add bx, 40000
add cx, 65446
```
instead of looking like this:
```
add bx, 40000
add cx, -90
```

I hope it makes sense, and it's not too much of a nitpicking. I just though that it could help somebody who is following my path of completely matching your listings outputs.

**Comment by cmuratori** [comment]
I guess I just don't understand the argument for printing everything unsigned. It's an arbitrary choice, and to me it seems more suitable to print 8-bit values as signed because then it is more obvious that they were 8 bits, not 16.

I agree that we _could_ print them as unsigned, because AFAIK assemblers will still correctly figure out that you can encode 65446 as an 8-bit sign-extended value. But to me it seems clearer to write -90, so that the programmer can immediately see that this was an 8-bit immediate, instead of 65446, where it _appears_ that it was a 16-bit immediate even though it actually wasn't.

\- Casey

**Comment by pankkor** [comment]
I see. In my head it's more like: we simulate 16-bit processor that's why we sign extend 1 byte to a 16-bit word.

P.S. I don't have access to x86 hardware at the moment and I can't check what would be that outcome of this program.
If we run `83 C1 A6` (`add cx, -90`) in 32-bit mode, would it be like adding 65446 or -90? I
x86 documentation says that:
> DATA-SX: 8-bit immediate value that is automatically sign-extended to 16-bits before use.

And x386 manual says that sign extension depends on destination:
> When an immediate byte value is added to a word or doubleword operand, the immediate value is first sign-extended to the size of the word or doubleword operand

<img width="564" alt="image" src="https://github.com/cmuratori/computer_enhance/assets/3875331/dc2fa4f7-70f6-47f7-9f9c-074ee46ea484">

So my guess is that `83 C1 A6` would act like adding 65446 (0xFFA6) and not like adding -90(0xFFFF FFA6)

**Comment by pankkor** [comment]
I can see that it all doesn't matter if we only simulate 16bit processor. I'll leave my simulator to sign extend up to a word boundary and will amend your simulation listings to my liking. Thanks for you input

**Comment by pankkor** [comment]
I tested with x86_64. In 64-bit mode `add cx, -90` is encoded as `66 83 C1 A6` and immediate is sign extended to 16-bit value.

After these instructions `rcx` would be set to `0xFFA6`
```
mov rcx, 0
add rx, -90
```
